qiime2/sdk/plugin_manager.py

```diff
@@ -38,29 +38,57 @@ def iter_entry_points(cls):
 
     # This class is a singleton as it is slow to create, represents the
     # state of a qiime2 installation, and is needed *everywhere*
-    def __new__(cls):
+    def __new__(cls, add_plugins=True):
         if cls.__instance is None:
             self = super().__new__(cls)
-            self._init()
+            self._init(add_plugins=add_plugins)
             cls.__instance = self
         return cls.__instance
 
-    def _init(self):
+    def _init(self, add_plugins):
         self.plugins = {}
+        self._plugin_by_id = {}
         self.semantic_types = {}
         self.transformers = collections.defaultdict(dict)
         self.formats = {}
         self.views = {}
         self.type_formats = []
 
-        # These are all dependent loops, each requires the loop above it to
-        # be completed.
-        for entry_point in self.iter_entry_points():
-            plugin = entry_point.load()
-            self.plugins[plugin.name] = plugin
-
-        for plugin in self.plugins.values():
-            self._integrate_plugin(plugin)
+        if add_plugins:
+            # These are all dependent loops, each requires the loop above it to
+            # be completed.
+            for entry_point in self.iter_entry_points():
+                project_name = entry_point.dist.project_name
+                package = entry_point.module_name.split('.')[0]
+                plugin = entry_point.load()
+
+                self.add_plugin(plugin, package, project_name)
+
+    def add_plugin(self, plugin, package=None, project_name=None):
+        self.plugins[plugin.name] = plugin
+        self._plugin_by_id[plugin.id] = plugin
+        if plugin.package is None:
+            plugin.package = package
+        if plugin.project_name is None:
+            plugin.project_name = project_name
```

Comment:
we need to check if these are still `None` after applying arguments, if so, we'll want to error and force the caller to pass them.

Reply:
philosophical question: do we worry about mutating the object we import, that is to say, do we ever want to share the same plugin object between plugin-managers? It sounds like nonsense, but we are ruling that future out by doing this.